src/ir/auth_logic/lowering_ast_datalog.cc

```diff
@@ -262,11 +273,129 @@ std::vector<datalog::Rule> LoweringToDatalogPass::QueriesToDLIR(
+  // (possibly nested) cansay expressions are based on an environment
+  // that tracks the maximum depth with which they appear. This environment is
+  // populated by side-effect while translating the program body. As a result,
+  // it is assumed that this function is called after the body is translated.
+
+  // The base relation declarations are the same as the ones from the
+  // surface program plus another one for "canActAs"
+  std::vector<datalog::RelationDeclaration> base_declarations =
+      relation_declarations;
+  base_declarations.push_back(datalog::RelationDeclaration(
+      "canActAs", false,
+      {datalog::Argument(
+           "p1", datalog::ArgumentType(datalog::ArgumentType::Kind::kPrincipal,
+                                       "Principal")),
+       datalog::Argument(
+           "p2", datalog::ArgumentType(datalog::ArgumentType::Kind::kPrincipal,
+                                       "Principal"))}));
+  // Attributes have a principal that the attribute is applied to
+  // as an additional parameter at the beginning of the param list.
+  // Transform the attributes
+  std::vector<datalog::RelationDeclaration> transformed_declarations =
+      utils::MapIter<datalog::RelationDeclaration>(
+          base_declarations, [](datalog::RelationDeclaration declaration) {
+            if (declaration.is_attribute()) {
+              std::vector<datalog::Argument> transformed_arguments = {
+                  datalog::Argument("attribute_prin",
+                                    datalog::ArgumentType(
+                                        datalog::ArgumentType::Kind::kPrincipal,
+                                        "Principal"))};
+              transformed_arguments.insert(transformed_arguments.end(),
+                                           declaration.arguments().begin(),
+                                           declaration.arguments().end());
+              return datalog::RelationDeclaration(declaration.relation_name(),
+                                                  false, transformed_arguments);
+            } else {
+              return declaration;
+            }
+          });
+  // Produce a mapping from predicate names to predicate typings
+  // (where a predicate typing is the same as a predicate relation declaration)
+  absl::flat_hash_map<std::string_view, datalog::RelationDeclaration>
+      type_environment;
+  for (datalog::RelationDeclaration declaration : transformed_declarations) {
+    type_environment.insert({declaration.relation_name(), declaration});
+  }
+
+  // Generate relation declarations for delegated predicates with
+  // each encountered delegation depth
+  std::vector<datalog::RelationDeclaration> can_say_declarations;
+  for (const auto& can_say_depth_pair : can_say_depth) {
+    auto type_iterator = type_environment.find(can_say_depth_pair.first);
+
+    if ((can_say_depth_pair.second > 0) &&
+        (type_iterator != type_environment.end())) {
```

Comment:
Negating the conditions might make it easier to read:
```suggestion
    if (depth == 0) continue;
    auto type_iterator = type_environment.find(type_name);
    if (type_iterator == type_environment.end() continue;
```